katsdpcontroller/product_config.py

```diff
@@ -31,63 +82,1196 @@ class _Sensor(ABC):
+            raise ValueError(
+                f'channels per endpoint ({self.n_channels_per_endpoint}) '
+                f'is not a multiple of channels per substream ({self.n_channels_per_substream})'
+            )
+
+    @property
+    def antenna_channelised_voltage(self) -> 'AntennaChannelisedVoltageStreamBase':
+        return cast(AntennaChannelisedVoltageStreamBase, self.src_streams[0])
+
+    @property
+    def antennas(self) -> Sequence[str]:
+        """Antenna names."""
+        return self.antenna_channelised_voltage.antennas
+
+    @property
+    def n_channels(self) -> int:
+        """Number of channels."""
+        return self.antenna_channelised_voltage.n_channels
+
+    @property
+    def n_channels_per_endpoint(self) -> int:
+        return self.n_channels // self.n_endpoints
+
+    @property
+    def n_substreams(self) -> int:
+        return self.n_channels // self.n_channels_per_substream
+
+    @property
+    def n_antennas(self) -> int:
+        """Number of antennas."""
+        return len(self.antennas)
+
+    @property
+    def bandwidth(self) -> float:
+        """Output bandwidth, in Hz."""
+        return self.antenna_channelised_voltage.bandwidth
+
+    @property
+    def centre_frequency(self) -> float:
+        """Sky centre frequency, in Hz."""
+        return self.antenna_channelised_voltage.centre_frequency
+
+    @property
+    def adc_sample_rate(self):
+        """ADC sample rate, in Hz."""
+        return self.antenna_channelised_voltage.adc_sample_rate
+
+    @property
+    def n_samples_between_spectra(self):
+        """Number of ADC samples between spectra."""
+        return self.antenna_channelised_voltage.n_samples_between_spectra
+
+    @property
+    @abstractmethod
+    def size(self) -> int:
+        """Size of a single frame in bytes."""
+
+    @property
+    @abstractmethod
+    def int_time(self) -> float:
+        """Time between heaps, in seconds."""
+
+    def net_bandwidth(self, ratio: float = 1.05, overhead: int = 128) -> float:
+        """Network bandwidth in bits per second."""
+        heap_size = self.size / self.n_substreams
+        return bandwidth(heap_size, self.int_time, ratio, overhead) * self.n_substreams
+
+
+class BaselineCorrelationProductsStreamBase(CbfPerChannelStream):
+    """Base for both simulated and real baseline-correlation-products streams."""
+
+    def __init__(self, name: str, src_streams: Sequence[Stream], *,
+                 int_time: float,
+                 n_endpoints: int,
+                 n_channels_per_substream: int,
+                 n_baselines: int,
+                 bits_per_sample: int):
+        super().__init__(
+            name, src_streams,
+            n_endpoints=n_endpoints,
+            n_channels_per_substream=n_channels_per_substream,
+            bits_per_sample=bits_per_sample
+        )
+        self._int_time = int_time
+        self.n_baselines = n_baselines
+
+    @property
```

Comment:
In this case it's because `int_time` is an abstract property of the base class. I don't want to make it an attribute of the base class because in TiedArrayChannelisedVoltageStream it's a function of the other parameters and cannot be set independently.